utils/include/utils/merge_sorted.h

```diff
@@ -0,0 +1,160 @@
+    }
+
+    const auto less = [&cmp](const auto& l, const auto& r) { return std::invoke(cmp, l, r) < 0; };
+    std::inplace_merge(sorted_vec.begin(), sorted_vec.begin() + init_size, sorted_vec.end(), less);
+}
+
+template<typename T,
+    typename R,
+    merge_key_comparator<std::vector<T>, decltype(std::begin(std::declval<R>()))> KeyComparator =
+        std::compare_three_way,
+    merge_updater<std::vector<T>, decltype(std::begin(std::declval<R>()))> Merger = trivial_merger>
+void merge_sorted(
+    std::vector<T>& sorted_vec, R&& sorted_range, KeyComparator&& cmp = KeyComparator(), Merger&& m = Merger())
+{
+    merge_sorted(sorted_vec,
+        std::begin(sorted_range),
+        std::end(sorted_range),
+        std::forward<KeyComparator>(cmp),
+        std::forward<Merger>(m));
+}
+
+template<typename T,
+    /* std::forward_iterator - move_iterator is viewed as input_iterator only before libc++17 (and possibly c++23) */
+    typename It,
+    std::sentinel_for<It> S,
+    merge_key_comparator<std::vector<T>, It> KeyComparator = std::compare_three_way,
+    merge_updater<std::vector<T>, It> Merger = trivial_merger>
+void merge_unsorted(
+    std::vector<T>& sorted_vec, It it, const S ite, KeyComparator&& cmp = KeyComparator(), Merger&& m = Merger())
+{
+    constexpr const auto trivial_inserter = !std::invocable<Merger, typename std::iterator_traits<It>::reference>;
+    auto&& p = [&m]() -> std::conditional_t<trivial_inserter, std::identity, Merger&> {
+        if constexpr (trivial_inserter)
+            return std::identity();
+        else
+            return m;
+    }();
+    const auto less = [&cmp](const auto& l, const auto& r) { return std::invoke(cmp, l, r) < 0; };
+    const auto init_size = sorted_vec.size();
+    for (; it != ite; ++it)
+    {
+        auto match = std::lower_bound(sorted_vec.begin(), sorted_vec.begin() + init_size, *it, less);
+        if (match != sorted_vec.begin() + init_size && std::invoke(cmp, *match, *it) == 0)
+            std::invoke(m, *match, *it);
+        else
+            sorted_vec.push_back(std::invoke(p, *it));
+    }
+    // libc++: std::views::transform(std::views::subrange(it, ite), std::ref(p))...
+    if constexpr (trivial_inserter)
+        sorted_vec.insert(sorted_vec.end(), it, ite);
+    else
+    {
+        for (; it != ite; ++it)
+            sorted_vec.push_back(std::invoke(p, *it));
+    }
+    std::stable_sort(sorted_vec.begin() + init_size, sorted_vec.end(), less);
+    std::inplace_merge(sorted_vec.begin(), sorted_vec.begin() + init_size, sorted_vec.end(), less);
+}
+
+template<typename T,
+    typename R,
+    merge_key_comparator<std::vector<T>, decltype(std::begin(std::declval<R>()))> KeyComparator =
+        std::compare_three_way,
+    merge_updater<std::vector<T>, decltype(std::begin(std::declval<R>()))> Merger = trivial_merger>
+void merge_unsorted(
+    std::vector<T>& sorted_vec, R&& unsorted_range, KeyComparator&& cmp = KeyComparator(), Merger&& m = Merger())
```

Comment:
Is it expected that `sorted_vec` is indeed sorted in the call to `merge_unsorted(...)`?

Reply:
Yes, the idea is to merge_sorted or merge_unsorted range into an already sorted one.